AlphaWallet/Tokens/Helpers/CachedERC1155ContractDictionary.swift

```diff
@@ -0,0 +1,76 @@
+//
+//  CachedERC1155ContractDictionary.swift
+//  AlphaWallet
+//
+//  Created by Jerome Chan on 12/4/22.
+//
+
+import Foundation
+
+class CachedERC1155ContractDictionary {
+    private var baseDictionary: [AlphaWallet.Address: Bool] = [AlphaWallet.Address: Bool]()
+    private var fileUrl: URL
+    private var encoder: JSONEncoder
+
+    static func loadFromCache(fileName: String) -> CachedERC1155ContractDictionary? {
+        do {
+            var url: URL = try FileManager.default.url(for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
+            url.appendPathComponent(fileName)
+            let cache: CachedERC1155ContractDictionary = CachedERC1155ContractDictionary(url: url)
+            if FileManager.default.fileExists(atPath: url.path) {
+                cache.readFromFileUrl()
+            }
+            return cache
+        } catch {
+            return nil
+        }
+    }
+
+    private init(url: URL) {
```

Comment:
What's the rationale to write a separate factory method instead of putting it into `init()` like this:

```
init?(fileName: String) {
    do {
        var url: URL = try FileManager.default.url(for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
        url.appendPathComponent(fileName)

        self.fileUrl = url
        self.encoder = JSONEncoder()

        if FileManager.default.fileExists(atPath: url.path) {
            readFromFileUrl()
        }
    } catch {
        return nil
    }
}
```

I didn't check if the code above works correctly.

Reply:
Reveals intent? I think I prefer functions and initialisers to be just simple and do one thing?

Reply:
This logic is sound generally, but apply this logic for initializers, wouldn't all initializers have to be private and quite often properties would have to be made `var` instead of `let` just because they are initialized in the factory method but otherwise not modified elsewhere? For that reason (var vs let) alone, it seems worth doing it in the `init` most of the time.

Reply:
No since this instance we need to use a cache file in a specific area. If you want me to do it in the init way, tell me and I'll do it otherwise I'd leave it like this. @hboon

Reply:
init way.